### PR TITLE
calorie related math effect

### PR DESCRIPTION
**Pull request by GuardianDll**
#### Summary
None
#### Purpose of change
Tried to add a way for math to read item calorie content, decided the best approach would be moving entire dialogue condition to the math
#### Describe the solution
make a math function `calories()`, that contain `stored_kcal` (as `calories()` or `calories('format': 'raw')`), stored_kcal_percentage (as `calories('format': 'percent')`) and also can read the caloric content of the item, if one of the talkers is item
#### Testing
image
image
#### Additional context
~~Albeit wonky, it worked before, but then i was told if i want optional options, i need to use kwargs instead of params, and when i changed it, it stopped working with items, which was kinda the point of all of it~~ it worked, i just didn't use the proper EoC, duh

### CODE REVIEW

File: src/math_parser_diag.cpp [hunk 1]
```diff
@@ -1317,6 +1317,56 @@ std::function<double( dialogue & )> vision_range_eval( char scope,
...
+            }
+            item_location *it = d.actor( beta )->get_item();
+            if( it && *it ) {
+                npc dummy;
+                return dummy.compute_effective_nutrients( *it->get_item() ).kcal();
+            }
+        } else {
+            debugmsg( "For calories(), talker is not character nor item" );
+            return 0;
+        };
```
Comment (mqrause) [inline src/math_parser_diag.cpp:1359]:
```suggestion
        }
        
        debugmsg( "For calories(), talker is not character nor item" );
        return 0;
```

This should fix the issue. If it's wrapped in an `else`, it won't catch paths like when `if( it && *it )` was false.

File: src/math_parser_diag.cpp [hunk 2]
```diff
@@ -1321,6 +1321,55 @@ std::function<double( dialogue & )> vision_range_eval( char scope,
...
+            debugmsg( R"(Unknown format type "%s" for calories, assumning "raw")", format );
+            format = "raw";
+        }
+
+        if( format == "percent" ) {
+            if( d.actor( beta )->get_character() ) {
+                int divisor = d.actor( beta )->get_healthy_kcal() / 100;
+                //if no data, default to default height of 175cm
+                if( divisor == 0 ) {
+                    divisor = 118169 / 100;
```
Comment (anothersimulacrum) [inline src/math_parser_diag.cpp:1344]: Where does the 118169 magic number come from? Either use the calculation that generates it or just have a debugmsg here and give some safe value, not a (liable to change) calculated value.

File: src/math_parser_diag.cpp [hunk 3]
```diff
@@ -1321,6 +1321,55 @@ std::function<double( dialogue & )> vision_range_eval( char scope,
...
+    return[format_value, beta = is_beta( scope )]( dialogue const & d ) {
+        std::string format = format_value.str( d );
+        if( format != "raw" && format != "percent" ) {
+            debugmsg( R"(Unknown format type "%s" for calories, assumning "raw")", format );
+            format = "raw";
+        }
+
+        if( format == "percent" ) {
+            if( d.actor( beta )->get_character() ) {
+                int divisor = d.actor( beta )->get_healthy_kcal() / 100;
```
Comment (anothersimulacrum) [inline src/math_parser_diag.cpp:1341]: Do floating point math with this, all EOC math is.
```suggestion
                double divisor = d.actor( beta )->get_healthy_kcal() / 100.0;
```

File: src/math_parser_diag.cpp (hunk 2 above)
Comment (GuardianDll) [reply src/math_parser_diag.cpp:1344]: right from the original `stored_kcal_percentage` implementation
https://github.com/CleverRaven/Cataclysm-DDA/pull/73136/files#diff-534f433e7a0b5ec5c233390037e131f3717e8aa5f347c47e650a65a1ddc95279L2133

Comment (anothersimulacrum) [reply src/math_parser_diag.cpp:1344]: Sorry, that was a rhetorical question. There shouldn't be a magic number like this.

File: src/math_parser_diag.cpp [hunk 4]
```diff
@@ -1321,6 +1321,59 @@ std::function<double( dialogue & )> vision_range_eval( char scope,
...
+        if( format != "raw" && format != "percent" )
+        {
+            debugmsg( R"(Unknown format type "%s" for calories, assumning "raw")", format );
+            format = "raw";
+        }
+
+        if( format == "percent" )
+        {
+            if( d.actor( beta )->get_character() ) {
+                int divisor = d.actor( beta )->get_healthy_kcal() / 100.0;
```
Comment (anothersimulacrum) [inline src/math_parser_diag.cpp:1343]:
```suggestion
                double divisor = d.actor( beta )->get_healthy_kcal() / 100.0;
```

Comment (GuardianDll) [reply src/math_parser_diag.cpp]: right, it needs to be turned back now

**Pull request merged**